Add Formatting Guide to GitHub (#3097)
opensearch-project/documentation-website: FORMATTING_GUIDE.md

# Formatting Guide

This guide provides an overview of the formatted elements commonly used in the OpenSearch documentation. 

* * *

> ### Table of contents

Reviewer: Nice touch!

* [Text style](#text-style)
* [Dashes](#dashes)
* [Horizontal rule](#horizontal-rule)
* [Unordered lists](#unordered-lists)
* [Ordered lists](#ordered-lists)
* [Nested lists](#nested-lists)
* [Tables](#tables)
* [Images](#images)
* [Labels](#labels)
* [Callouts](#callouts)
* [Buttons](#buttons)

> * [Adding pages or secitons](#adding-pages-or-sections)

Reviewer: sections

**Example**: 

```
In case of a cluster or node failure, all PIT data is lost.
{: .note}
```

For a callout with multiple paragraphs or lists, use `>`:

```
>   ****PREREQUISITE****
>
>   To use a custom vector map with GeoJSON, install these two required plugins:
>   * OpenSearch Dashboards Maps [`dashboards-maps`](https://github.com/opensearch-project/dashboards-maps_) front-end plugin
>   * OpenSearch [`geospatial`](https://github.com/opensearch-project/geospatial_) backend plugin
{: .note}

```

## Buttons

You can use either `copy` or `copy-curl` includes for code snippets. The `copy` include places a **Copy** button on the code snippet, while the `copy-curl` include places both **Copy** and **Copy as cURL** buttons. Use the `copy-curl` include for API requests. If an API request is already in the cURL format, use the `copy` include.

**Example of a `copy` include:**

````
```bash
curl -XGET "localhost:9200/_tasks?actions=*search&detailed
```
{% include copy.html %}
````

**Example of a `copy-curl` include:**

````
```json
PUT /sample-index1/_clone/cloned-index1
{
  "aliases": {
    "sample-alias1": {}
  }
}
```
{% include copy-curl.html %}
````

## Adding pages or sections

> When adding a page or a section, make the `nav_order` of the child pages multiples of 10. For example, if you have a parent page `Clients`, make child pages `Java`, `Python`, and `JavaScript` have a `nav_order` of 10, 20, and 30, respectively. Doing so makes inserting child pages easier.

Reviewer 2: should we say something like "adding additional child pages..." or something as the reasoning as well. Similar to the way BASIC is coded is my thought.

## Videos

To insert a video, add a youtube player include similar to the following:

```
{% include youtube-player.html id='_g46WiGPhFs' %}
```

> Note that the `id` variable refers to the youtube video ID that is the last part of the URL. For example, the youtube video at the URL `https://youtu.be/_g46WiGPhFs` has the id `_g46WiGPhFs`. The ID must be surrounded with single quotation marks.

Reviewer 2: Sounds a little off to me at least. Maybe like "the `id` variable refers to the youtube video ID at the end of the URL" ?

## Math

> To add mathematical expressions to a page, add `has_math: true` to the page’s front matter.

Reviewer 2: Just to clarify so I understand, when we say "front matter" are we referring to the header area of the md file?

Author: Yes, correct

**Example with code snippets**:

```
1. Run the demo batch script.
   There are two ways of running the batch script:
   1. Run the batch script using the Windows UI:
      1. Navigate to the top directory of your OpenSearch installation and open the `opensearch-{{site.opensearch_version}}` folder.
      1. Run the batch script by double-clicking the `opensearch-windows-install.bat` file. This opens a command prompt with an OpenSearch instance running.
   1. Run the batch script from Command prompt or Powershell:
      1. Open Command Prompt by entering `cmd`, or Powershell by entering `powershell`, in the search box next to ****Start**** on the taskbar. 
      1. Change to the top directory of your OpenSearch installation.
         ```bat
         cd \path\to\opensearch-{{site.opensearch_version}}
         ```
      1. Run the batch script.
         ```bat
         .\opensearch-windows-install.bat
         ```
```

**Example with images**:

```
1. To begin, select the rule in the Rule name column. The rule details pane opens.
    <img src="{{site.url}}{{site.baseurl}}/images/Security/rule-dup2.png" alt="Opening the rule details pane" width="50%">

> 1. Select the **Duplicate** button in the upper-right corner of the pane. The Duplicate rule window opens in Visual Editor view and all of the fields are automatically populated with the rule's details. Details are also populated in YAML Editor view.

Reviewer 3: For both of these examples, can we insert "..., as shown in the following image" in order to reinforce that style standard?

    <img src="{{site.url}}{{site.baseurl}}/images/Security/dupe-rule.png" alt="Selecting the duplicate button opens the Duplicate rule window" width="50%">
```

## Tables

Markdown table columns are automatically sized, and there is no need to specify a different number of dashes in the formatting. 

> **Example**:

Reviewer 3: ```suggestion
**Example**
```

```
Header 1 | Header 2
:--- | :---

> Body 1 | Body 2 that is extremely lengthy but there is no need to specify its width

Reviewer 3: ```suggestion
Body 1 | Body 2, which is extremely lengthy, but there is no need to specify its width.
```

Markdown images are responsive by default. To insert a Markdown image, use the `![<alternate text>](link)` syntax:

```
![OS branding]({{site.url}}{{site.baseurl}}/images/brand.png)
```

Markdown uses the image’s actual width to render it. It sets the maximum image width to the width of the main body panel.

If you want to specify the image width or another style, use HTML syntax:

```
<img src="{{site.url}}{{site.baseurl}}/images/brand.png" alt="OS branding" width="700"/>
```

You can specify width as a hard-coded number of pixels, as in the preceding example, or as a percentage of the parent width:

```
<img src="{{site.url}}{{site.baseurl}}/images/brand.png" alt="OS branding" width="70%"/>
```

To stretch the image to fit the width of the main body panel, use width=“100%”.

To take high-resolution screenshots, in Firefox, right-click on the page and choose “Take Screenshot”.

Image borders are automatic; do not manually add a border to an image.

Always **separate an image from the text with a blank line**:

```

> To send a query to OpenSearch, select the query by placing the cursor anywhere in the query text. Then choose the triangle on the top right of the request or press `Ctrl/Cmd+Enter`:

Reviewer 3: Is "press" the right word here?

Author: It is now that it's in the style guide :)

<img src="{{site.url}}{{site.baseurl}}/images/dev-tools/dev-tools-send.png" alt="Send request">
```

Do not place an image next to text or insert artificial line breaks using `<br>`. Otherwise, the text might render as aligned to the bottom of the image, with the image on the right. 

If the image is under a list item, place it on a new line with a tab. For more examples, see [Lists with code snippets or images](#lists-with-code-snippets-or-images).

## Labels

You can use the following labels:

* label-blue 
* label-green
* label-purple 
* label-red 
* label-yellow 

> Use a purple label to specify in which version an API was introduced:

Reviewer 3: ```suggestion
Use a purple label to specify the version in which an API was introduced:
```

```
# Alias
Introduced 1.0
{: .label .label-purple }
```

If we introduce a breaking change to an operation, add an additional label with a link to the release note for that breaking change:

```
## Get roles
Introduced 1.0
{: .label .label-purple }
[Last breaking change 2.0](https://example.com)
{: .label .label-red }
```

## Callouts

You can use four levels of callouts:

*  `{: .note}` blue
*  `{: .tip }` green
*  `{: .important}` yellow
*  `{: .warning}` red

Place a callout directly under the paragraph to which you want to apply the callout style.

> **Example**: 

Reviewer 3: ```suggestion
**Example**
```

```
In case of a cluster or node failure, all PIT data is lost.
{: .note}
```

For a callout with multiple paragraphs or lists, use `>`:

```
>   ****PREREQUISITE****
>
>   To use a custom vector map with GeoJSON, install these two required plugins:
>   * OpenSearch Dashboards Maps [`dashboards-maps`](https://github.com/opensearch-project/dashboards-maps_) front-end plugin
>   * OpenSearch [`geospatial`](https://github.com/opensearch-project/geospatial_) backend plugin
{: .note}

```

## Buttons

You can use either `copy` or `copy-curl` includes for code snippets. The `copy` include places a **Copy** button on the code snippet, while the `copy-curl` include places both **Copy** and **Copy as cURL** buttons. Use the `copy-curl` include for API requests. If an API request is already in the cURL format, use the `copy` include.

> **Example of a `copy` include:**

Reviewer 3: ```suggestion
**Example of a `copy` include**
```

````
```bash
curl -XGET "localhost:9200/_tasks?actions=*search&detailed
```
{% include copy.html %}
````

> **Example of a `copy-curl` include:**

Reviewer 3: ```suggestion
**Example of a `copy-curl` include**
```

````
```json
PUT /sample-index1/_clone/cloned-index1
{
  "aliases": {
    "sample-alias1": {}
  }
}
```
{% include copy-curl.html %}
````

## Adding pages or sections

> When adding a page or a section, make the `nav_order` of the child pages multiples of 10. For example, if you have a parent page `Clients`, make child pages `Java`, `Python`, and `JavaScript` have a `nav_order` of 10, 20, and 30, respectively. Doing so makes inserting additional child pages easier because it does not require renumbering existing pages.

Reviewer 3: ```suggestion
When adding a page or a section, make the `nav_order` of the child pages multiples of 10. For example, if you have a parent page `Clients`, make child pages `Java`, `Python`, and `JavaScript` have a `nav_order` of 10, 20, and 30, respectively. Doing so makes inserting additional child pages easier because it does not require you to renumber existing pages.
```

## Videos

> To insert a video, add a youtube player include similar to the following:

Reviewer 3: ```suggestion
To insert a video, add a YouTube player include similar to the following:
```

```
{% include youtube-player.html id='_g46WiGPhFs' %}
```

> Note that the `id` variable refers to the youtube video ID at the end of the URL. For example, the youtube video at the URL `https://youtu.be/_g46WiGPhFs` has the id `_g46WiGPhFs`. The ID must be surrounded with single quotation marks.

Reviewer 3: ```suggestion
Note that the `id` variable refers to the YouTube video ID at the end of the URL. For example, the YouTube video at the URL `https://youtu.be/_g46WiGPhFs` has the ID `_g46WiGPhFs`. The ID must be surrounded with single quotation marks.
```

```
## Math

Some Markdown paragraph. Here's a formula:

<p>
  When \(a \ne 0\), there are two solutions to \(ax^2 + bx + c = 0\) and they are
  \[x = {-b \pm \sqrt{b^2-4ac} \over 2a}.\]
</p>

And back to Markdown.
```

## Variables in curly braces

To correctly display variables that are in curly braces, escape the curly braces with the `{% raw %}{% endraw %}` tags:

````
"message_template": {
    "source": "the index is {% raw %}{{ctx.index}}{% endraw %}"
}
````

The variable `ctx.index` is rendered in double curly braces.

## Links

To add a link to a document, section, or image, use the `[name](link)` syntax, for example:

```
## Looking for the Javadoc?

See [opensearch.org/javadocs/](https://opensearch.org/javadocs/).
```

### Section links

**Section links** are links to headings in your document. Markdown lowercases the headings for links, drops back ticks, and replaces spaces with hyphens:

```
## The `minimum_should_match` parameter

For more information, see [the `minimum_should_match` parameter](#the-minimum_should_match-parameter).
```

### Internal links

> **Internal links** are links to another document or image within the documentation website. Because the documentation website is versioned, do not hard code the version number in the link. Use the relative path where `{{site.url}}{{site.baseurl}}` refers to the main directory instead:

Reviewer 3: Should the last sentence have commas like so?: "Use the relative path, where `{{site.url}}{{site.baseurl}}` refers to the main directory, instead:"
